src/hotspot/share/opto/predicates.hpp

```diff
@@ -416,23 +542,207 @@ class InitializedAssertionPredicate : public StackObj {
   IfTrueNode* create_success_path(IfNode* if_node, IdealLoopTree* loop);
 };
 
+// This class iterates through all predicates of a Regular Predicate Block and applies the given visitor to each.
+class RegularPredicateBlockIterator : public StackObj {
+  Node* const _start_node;
+  const Deoptimization::DeoptReason _deopt_reason;
+
+ public:
+  RegularPredicateBlockIterator(Node* start_node, Deoptimization::DeoptReason deopt_reason)
+      : _start_node(start_node),
+        _deopt_reason(deopt_reason) {}
+  NONCOPYABLE(RegularPredicateBlockIterator);
+
+  // Skip all predicates by just following the inputs. We do not call any user provided visitor.
+  Node* skip_all() const {
+    PredicateVisitor do_nothing; // No real visits, just do nothing.
+    return for_each(do_nothing);
+  }
+
+  // Walk over all predicates of this block (if any) and apply the given 'predicate_visitor' to each predicate.
+  // Returns the entry to the earliest predicate.
+  Node* for_each(PredicateVisitor& predicate_visitor) const {
```

Comment:
The goal is eventually to only have this method which does the check for what kind of predicate we face in the graph and replace all other places where we do these kind of checks together with visitors.